### PR TITLE
client/Txn: support CommitBy with a deadline

### LINKED ISSUE #2820: support a MaxTimestamp parameter to EndTransaction
In order to support table leases (#2810) we need to have a guarantee that a transaction does not commit after the leases for any tables referenced in the transaction expire. We need the KV transaction to support a "write fence": a timestamp after which a committed transaction will be aborted instead of committed. This could be accomplished by adding a `MaxTimestamp` parameter to `EndTransaction` requests. If the proposed commit timestamp is larger than the `MaxTimestamp` the transaction would be aborted.

### PR DESCRIPTION
**Pull request by tamird**
Fixes #2820.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3028)

### CODE REVIEW

File: kv/txn_coord_sender_test.go [hunk 1]
```diff
@@ -317,15 +318,59 @@ func TestTxnCoordSenderEndTxn(t *testing.T) {
...
+					t.Error(err)
+				}
+			case 3:
+				// future deadline
+				if _, ok := err.(*roachpb.TransactionAbortedError); !ok {
+					t.Errorf("expected TransactionAbortedError but got %T: %s", err, err)
+				}
+			}
+		}
+		verifyCleanup(key, s.Sender, s.Eng, t)
```
Comment (tamird) [inline kv/txn_coord_sender_test.go]: @tschottdorf this fails for `case 3:`, but I'm not sure why. Any insight?

**Comment by tschottdorf** [comment]
Reviewed 9 of 9 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[kv/txn_coord_sender_test.go, line 307 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMBYOq757ZFATz5HJ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L307)):</sup>
s/expected empty metadata/leftover write intent/


---

<sup>**[kv/txn_coord_sender_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMekPd1VRRGSlNazq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L321)):</sup>
nit: s/$/./


---

<sup>**[kv/txn_coord_sender_test.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMlliOoCIQwylDO37)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L351)):</sup>
nit: capitalization and dot


---

<sup>**[kv/txn_coord_sender_test.go, line 372 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2T-hkuHzQklX4p7_0b-r1-372)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L372)):</sup>
yep, see my other comment in EndTransaction.


---

<sup>**[storage/replica_command.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UOpznPI3uvmyqtcAg)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/storage/replica_command.go#L461)):</sup>
you're returning an error, and that means that whatever you've written to the batch is actually not executed, in particular the resolution of your local intents is tossed. I should've remembered that this was an issue (notice that the other TransactionAbortedError above doesn't resolve its local intents here but passes them to the caller, for precisely that reason). There are some options:

a) you return all intents here, not only externalIntents, but you need to make sure that they all have the transaction as ABORTED. That way they're going to remove the intents, but you have no way of actually writing ABORTED in the master proto. That's super dirty and there might be an anomaly hidden somewhere.
b) if the commit only fails because of the deadline, you return this information not as an error but encoded in the response. This is similar to the recent discussions about ConditionalPut. A client would not receive a TransactionAbortedError but instead find out that its proto isn't PENDING any more. Semantically that should actually work out well, but needs a bunch of refactoring (TransactionAbortedError would just go away, we don't really need it - a BatchResponse carries a transaction proto as well).
c) we break with the invariant that a command that errors out can't write to the engine. See https://github.com/cockroachdb/cockroach/issues/1989

I personally think the prettiest option is dropping TransactionAbortedError altogether, though it's late on a Friday and I might be missing something.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tamird** [comment]
Review status: 6 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

<sup>**[kv/txn_coord_sender_test.go, line 307 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMBYOq757ZFATz5HJ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L307)):</sup>
Can you help me understand why that would be the correct error message? Should I just check for `ok` and not bother with `meta.Txn != nil`?


---

<sup>**[kv/txn_coord_sender_test.go, line 321 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMekPd1VRRGSlNazq)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L321)):</sup>
Done.


---

<sup>**[kv/txn_coord_sender_test.go, line 351 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMlliOoCIQwylDO37)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L351)):</sup>
Done.


---

<sup>**[storage/replica_command.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UOpznPI3uvmyqtcAg)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/storage/replica_command.go#L461)):</sup>
Done. Went with (a) for now.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[kv/txn_coord_sender_test.go, line 307 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMBYOq757ZFATz5HJ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L307)):</sup>
no, the check is fine. You're checking that there isn't an intent (so either no value, or a real value). If you find an intent, it's a leftover write intent and you error out.


---

<sup>**[storage/replica_command.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UOpznPI3uvmyqtcAg)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/storage/replica_command.go#L461)):</sup>
Works for me. Please link the issue you're filing to this PR and put a TODO in place where you set the txn to ABORTED above.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

File: storage/replica_test.go [hunk 2]
```diff
@@ -1588,6 +1589,69 @@ func TestRangeResponseCacheStoredError(t *testing.T) {
 	}
 }
 
+// TestEndTransactionDeadline verifies that EndTransaction respects the
+// transaction deadline.
+func TestEndTransactionDeadline(t *testing.T) {
```
Comment (tschottdorf) [inline storage/replica_test.go]: isn't the corresponding test in `kv` just a one-to-one copy of this one? Is there a reason for keeping both?

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[kv/txn_coord_sender_test.go, line 307 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UMBYOq757ZFATz5HJ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/kv/txn_coord_sender_test.go#L307)):</sup>
Done.


---

<sup>**[storage/replica_command.go, line 461 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2UOpznPI3uvmyqtcAg)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/storage/replica_command.go#L461)):</sup>
Done. Just pushed with a comment.


---

<sup>**[storage/replica_test.go, line 1594 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2T-hkuHzQklX4p7_0h-r4-1594)** ([raw file](https://github.com/cockroachdb/cockroach/blob/bc954b9f68ae78e2b4da790a0aed27a1f6b49045/storage/replica_test.go#L1594)):</sup>
Yes, but this one managed to pass while the `kv` one didn't. Would you prefer I removed this one?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

<sup>**[storage/replica_test.go, line 1594 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2T-hkuHzQklX4p7_0h-r4-1594)** ([raw file](https://github.com/cockroachdb/cockroach/blob/bc954b9f68ae78e2b4da790a0aed27a1f6b49045/storage/replica_test.go#L1594)):</sup>
the kv one verifies that the intent is resolved, this one doesn't. I think we can keep both for now, it's not like they take a long time.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

**Comment by tschottdorf** [comment]
LGTM

**Comment by petermattis** [comment]
LGTM


---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[client/txn.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2YZSzokerYO-NVYeO0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/client/txn.go#L340)):</sup>
This could call `txn.CommitBy(nil)`, though for only saving 2 lines it doesn't seem particularly important to change now. 


---

<sup>**[client/txn.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028#-K2YZbqF2Cto07hEnY-_)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3e71918e692ca88f26a5019ec52dbedc80660e04/client/txn.go#L344)):</sup>
Might want to call out that passing `deadline == nil` is equivalent to no deadline.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3028)

File: kv/txn_coord_sender_test.go [hunk 3]
```diff
@@ -317,15 +318,59 @@ func TestTxnCoordSenderEndTxn(t *testing.T) {
...
+				if err != nil {
+					t.Error(err)
+				}
+			case 2:
+				// Equal deadline.
+				if err != nil {
+					t.Error(err)
+				}
+			case 3:
+				// Future deadline.
```
Comment (bdarnell) [inline kv/txn_coord_sender_test.go]: Isn't this backwards? Future deadlines should allow for a commit and past deadlines should fail.

**Pull request merged**